Sources/AST/Declaration/FunctionSignatureDeclaration.swift

```diff
@@ -0,0 +1,88 @@
+//
+//  FunctionSignatureDeclaration.swift
+//  AST
+//
+//  Created by Harkness, Alexander L on 2018-09-05.
+//
+import Source
+import Lexer
+
+// The declaration of a function signature.
+public struct FunctionSignatureDeclaration: ASTNode {
+  public var funcToken: Token
+
+  /// The attributes associated with the function, such as `@payable`.
+  public var attributes: [Attribute]
+
+  /// The modifiers associted with the function, such as `public` or `mutating.`
+  public var modifiers: [Token]
+  public var identifier: Identifier
+  public var parameters: [Parameter]
+  public var closeBracketToken: Token
+  public var resultType: Type?
+
+  public var mangledIdentifier: String? = nil
+
```

Comment:
Feels like we should generalise Function Heads rather than having all this code duplication

Reply:
That would probably be a good idea, I just copied to simplify initial work, I will see about de-duplicating.